Adding rich text block support

GetOutOfMyBakery opened the pull request:
# What
Adding support for [Rich Text Blocks](https://api.slack.com/reference/block-kit/blocks#rich_text):
- resolves: https://github.com/CGA1123/slack-ruby-block-kit/issues/184

Rich Text Blocks are a bit unique compared to other blocks:
> It is also the output of the Slack client's WYSIWYG message composer, so all messages sent by end-users will have this format. Use this block to include user-defined formatted text in your Block Kit payload. While it is possible to format text with `mrkdwn`, `rich_text` is strongly preferred and allows greater flexibility.

Most of the other blocks have separate "compatible block elements" and tend to be relatively shallow, whereas Rich Text Blocks can be "[deeply nested](https://api.slack.com/reference/block-kit/blocks#rich_fields)" and it's elements have no reuse outside of a `rich_text` block.

Also added in this PR:
- `./examples/`: scripts to generate example message that can either be:
  - directly posted to to Slack
  - copied and pasted to Slack's [Block Builder Kit](https://app.slack.com/block-kit-builder) UI to validate the output

This makes it much easier to build confidence in changes being made to the Gem by semi-automating validation against Slack's schema. Since it looks like there's no future opportunity to validate against the JSON schema automatically:
- https://github.com/slackapi/slack-api-specs/issues/18 (and the repo is now archived)

this strategy might be the best solution going forward.

CGA1123 commented:
Hey @GetOutOfMyBakery -- thanks for taking this on 🙇‍♂️

Let me know when you're happy for a review and I'll get to it as soon as possible.

\cc @dmorgan-fa -- would be keen to hear your feedback on this from, if you're able to begin integrating agains this branch for your use-case.

dmorgan-fa commented:
👋 Hey @CGA1123, that's my personal GitHub account and this is my work one 😄 

I started on it last night and got into a flow, so it's in pretty good shape just now, but I've not had a chance to validate against Slack's Block Kit Builder yet, since I didn't have access to a workspace from my personal machine.

On an unrelated note, I really appreciated how easy it was to work with your framework! It's really well architected and made it super easy to work on 🙏

CGA1123 commented:
Ha, in that case syncing up between the both of you should be a breeeeeze! 😄

Glad to hear it! It is a relatively simple (if not quite repetitive) flow, but it's quite cathartic working in some plain old ruby as a change of pace, with a fast test suite to go with it!

---

Will take a look later this evening and have play around with these new block on a Slack workspace 👍

GetOutOfMyBakery commented on examples/comprehensive_rich_text_blocks.rb:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env ruby
```
I've taken another look at `./examples/` and expanded it a bit:
- `config.rb`: for some configurable settings, specifically IDs that cause Slack Block Kit Builder to report as an error if they're invalid
- `simple_example.rb` broadly duplicating the example from the README
- `example_helper.rb` takes the `Config.webhook_url` (if present) and posts the message from the example script to Slack (I haven't tested this yet, but it's modelled on the README example)

I'd been doing a lot of `bundle exec example/<file>.rb | pbcopy` and pasting that directly into Slack Block Kit Builder to validate the output, so posting it directly to Slack could speed that cycle up.

GetOutOfMyBakery commented:
I'm happy to do another pass and tidy up the commit history, etc., later but functionality-wise it's looking good. Let me know if there's anything that jumps out at you that you don't like, or if you want to give some pointers about what you want updated before getting it ready to merge.

***

Slightly off-topic but there's a few places that Slack mention some hard limits:
> You can include up to 50 blocks in each message [...]

[rich_text](https://api.slack.com/reference/block-kit/blocks#rich_text)

> Minimum length for the text in this field is 1 and maximum length is 3000 characters.

[section_fields](https://api.slack.com/reference/block-kit/blocks#section_fields), but it applies to `rich_text` too

It's a small tripping hazard if you're composing complicated messages with lost of rich text elements, or if you're creating single long messages it can be a bit confusing initially.

Is this something you'd be up for gaurding against in the gem, or would you prefer to let Slack respond with the error messaging when limits are exceeded?

GetOutOfMyBakery commented on Gemfile:
```diff
@@ -12,6 +12,9 @@ gem 'rspec', '~> 3'
 gem 'rspec_junit_formatter', '~> 0.6'
 
 gem 'codecov', '~> 0.6', require: false
+gem 'debug', '~> 1.9', require: false
+gem 'racc', '~> 1.7', require: false
+gem 'retest', '~> 1.13', require: false
```
For transparency:
- `debug` isn't necessary, it's just a nice alternative/complement to `pry`
- `racc` was required for me to run `rubocop` (Ruby 3.3.0, Mac running Sonoma)
- [`retest`](https://github.com/AlexB52/retest) is just super helpful, essentially opinionated `guard` without a Guardfile, and it supports `focus` out of the box 😍

GetOutOfMyBakery commented on README.md:
```diff
@@ -67,8 +67,8 @@ blocks = Slack::BlockKit.blocks do |b|
   b.append(a_prebuilt_block)
 end
 
+body = { blocks: blocks.as_json }
```
I've dropped `text: 'New block message!'` because Slack Block Kit Builder UI throws an error when the line is present.

GetOutOfMyBakery commented on README.md:
```diff
@@ -81,7 +81,9 @@ This will create a message like this:
 ![example block message](https://git.io/fjDWR)
 
-You can also check out the [`slackerduty`](https://github.com/CGA1123/slackerduty) project for some example,
+See [`./examples`](./examples/) for more worked examples.
```
Planning to add a bit more guidance here with regard to updating the `config.rb`, etc.

CGA1123 commented:
> Is this something you'd be up for gaurding against in the gem, or would you prefer to let Slack respond with the error messaging when limits are exceeded?

This is very much in tune with what was wanted in #176 -- I think that would be quite useful to have, I wonder what the "right" behaviour is in these circumstances, truncations, erroring, some other way of handling it. I think for the time being (in the scope of this PR) we can leave this functionality out. When implemented it would only be useful if consistent across the library I think.

CGA1123 commented on lib/slack/block_kit/layout/rich_text/rich_text_elements.rb:
```diff
@@ -0,0 +1,36 @@
+            append(RichTextElements::Emoji.new(name: name))
+          end
+
+          def link(url:, text: nil, unsafe: false, styles: [])
```
```suggestion
          def link(url:, text: nil, unsafe: nil, styles: [])
```

CGA1123 reviewed (changes requested):
A couple of minor changes to keep behaviour as similar as the Slack API as possible!

FYI -- have dropped ruby 2.6 from the matrix (as EOL) and added 3.3 which required the addition of `racc` in the Gemfile which is now on `master` already. Quick rebase should sort bits out and make the build 🟢

Looking fantastic. Appreciate the time taken to work through adding this feature 🙇‍♂️

There are some minor gaps on the spec coverage, although nothing major. Might be nice to have a shared example for `RichText::RichTextElements` to include into the container objects if time allows -- non-blocking though from my perspective. 👍

GetOutOfMyBakery commented on lib/slack/block_kit/layout/rich_text/rich_text_elements/link.rb:
```diff
@@ -0,0 +1,36 @@
+            TYPE = 'link'
+            VALID_STYLES = %w[bold italic strike code].freeze
+
+            def initialize(url:, text: nil, unsafe: false, styles: [])
```
Good spot, I've sorted it now 👍

GetOutOfMyBakery commented on lib/slack/block_kit/layout/rich_text/rich_text_elements/style_helper.rb:
```diff
@@ -0,0 +1,27 @@
+
+              style_block = {}
+              styles.each { |style| style_block[style.to_s] = true }
+              style_block.compact
```
Ah, that's much tidier 👌

GetOutOfMyBakery commented:
@CGA1123, no need to apologise, you've been super responsive, don't sweat it.

I've rebased and taken on board your suggestions, take a look from commit [Applying Christian's suggested changes _4949baf_](https://github.com/CGA1123/slack-ruby-block-kit/pull/185/commits/4949baff88455c487758da2f9b263feab811fc62) onwards.

I've:
- updated the main `README.md`
- added a new `./examples/README.md` and updated `./examples/example_helper.rb`
- refactored `RichTextElements::StyleHelper` to make it a bit easier to work with
- increased spec coverage and used `shared_examples` where appropriate

I've not:
- tested posting directly to Slack with `./examples` scripts
- tidied up the commit history (happy to tidy it up or squash it once you're happy with the PR)

CGA1123 commented on examples/example_helper.rb:
```diff
@@ -0,0 +1,22 @@
+
+module ExampleHelper
+  def self.print_output(line)
+    puts line unless ENV.fetch('SUPPRESS_WEBHOOK_MESSAGE', nil) == 'true'
```
Could `$stderr.puts` this -- would also simplify the need for setting this env var when piping output!

```suggestion
    $stderr.puts line unless ENV.fetch('SUPPRESS_WEBHOOK_MESSAGE', nil) == 'true'
```

CGA1123 reviewed (approved):
Build failures are unrelated (codecov being unhappy for some reason), specs all running fine 👍 

I've tested out posting directly to slack and it's working a charm as best I can tell -- really like this addition of the `examples/` and related helpers (also TIL `[!TIP]` and `[!NOTE]`) 👏 

Quick squash of commits + an entry to `CHANGELOG` would be much appreciated and I'll get this merged and released. Nice one 🚀

GetOutOfMyBakery commented:
Thanks @CGA1123, great to have confirmation that the examples scripts are posting to Slack as expected 🙏

There's a few others: `[!IMPORTANT]`, `[!WARNING]`, `[!CAUTION]`. They're still relatively new [see here](https://github.com/orgs/community/discussions/16925).

I've squashed it all down, updated the `CHANGELOG`, and applied the `$stderr` change you'd suggested 👍

Thanks for the fast feedback and turn around on this 🙏

CGA1123 commented:
Thanks for contributing @GetOutOfMyBakery 🙇‍♂️ 

This has now shipped out in [`v0.24.0`](https://rubygems.org/gems/slack-ruby-block-kit/versions/0.24.0)
